File: LeetCode/Peng/151ReverseWordsInAString.cpp
/**
 * Use stack to re-construct string, OR use vector<pairs> to record the
 * words positions.
 **/ 
class Solution {
public:
    void reverseWords(string &s) {
        int n = s.size();
        stack<string> words;
        for (int i = 0; i < n; i++) {
            while (i < n && s.at(i) == ' ') {
                i++;
            }
            int pos = i;
            while (i < n && s.at(i) != ' ') {
                i++;
            }
            if (i - pos) {
                words.push(s.substr(pos, i - pos));
            }
        }
        string result = "";
        while (!words.empty()) {
            result += words.top();
            words.pop();
            while (!words.empty()) {
                result += " " + words.top();
                words.pop();
            }
        }
        s = result;
    }
};

File: LeetCode/Kai/151ReverseWordsInAString.java
public class Solution {
    public String reverseWords(String s) {
        s = s.trim();
        if(s.length() == 0) return "";
        String[] words = s.split("\\s+");
        StringBuilder sb = new StringBuilder();
        for(int i=words.length-1; i>=1; i--) {
            sb.append(words[i]);
            sb.append(" ");
        }
        sb.append(words[0]);
        
        return sb.toString();
    }
    
}